notebooks/examples/python/OpenStreetMaps/0_Download.py

```diff
@@ -0,0 +1,238 @@
+import pathlib
+
+# The DBFS file system is mounted under /dbfs/ directory on Databricks cluster nodes
+local_osm_path = pathlib.Path(raw_path.replace("dbfs:/", "/dbfs/"))
+
+(local_osm_path / "raw").mkdir(parents=True, exist_ok=True)
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC ## Explore available regions 
+
+# COMMAND ----------
+
+import requests
+
+resp = requests.get(mirror_index_url)
+features = resp.json()["features"]
+datasets = list(map(lambda x: x["properties"], features))
+
+available_regions = list(map(lambda x: f"{x['id']} ({ x.get('parent') })", datasets))
+available_regions
+
+# COMMAND ----------
+
+# Define which regions to download
+regions = [
+  'italy'
+  
+#   'africa', 
+#   'antarctica', 
+#   'asia', 
+#   'australia-oceania', 
+#   'central-america', 
+#   'europe',
+#   'north-america',
+#   'south-america'
+] # See available regions few cell above
+
+print(f"The raw Open Street Maps data for {regions} will be stored in {raw_path}")
+
+# COMMAND ----------
+
+# Uncomment the following line to clear all raw data
+# dbutils.fs.rm(raw_path + "raw", True)
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC 
+# MAGIC ## Download
+# MAGIC 
+# MAGIC Download the region files concurrently
+
+# COMMAND ----------
+
+from multiprocessing import Pool
+
+def download_region(region):
+  selected_dataset = list(filter(lambda x: x["id"] == region, datasets))[0]
+  compressed_osm_file = local_osm_path / "raw" / f"{region}.osm.bz2"
+  osm_url = selected_dataset["urls"]["bz2"]
+
+  # Download the file incrementally
+  with requests.get(osm_url, stream=True) as r:
+    r.raise_for_status()
+    with open(compressed_osm_file, 'wb') as f:
+      for chunk in r.iter_content(chunk_size=512*1024):
+        f.write(chunk)
+
+with Pool(4) as pool:
+  pool.map(download_region, regions)
+
+# COMMAND ----------
+
+display(dbutils.fs.ls(raw_path + "raw"))
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC ## Ingest bronze layer
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC ### Nodes
+
+# COMMAND ----------
+
+from pyspark.sql.types import *
+
+nodes_schema = StructType([
+    StructField("_id", LongType(), True),
+    StructField("_lat", DoubleType(), True),
+    StructField("_lon", DoubleType(), True),
+    StructField("_timestamp", TimestampType(), True),
+    StructField("_visible", BooleanType(), True),
+    StructField("_version", IntegerType(), True),
+    StructField("tag", ArrayType(
+      StructType([
+         StructField("_VALUE", StringType(), True),
+         StructField("_k", StringType(), True),
+         StructField("_v", StringType(), True),
+      ])
+    ), True)])
+
+nodes = (spark
+      .read
+      .format("xml")
+      .options(rowTag="node") # Only extract nodes
+      .load(f"{raw_path}/raw/", schema = nodes_schema)
+     )
+
+nodes.write.format("delta").mode("overwrite").save(f"{raw_path}/bronze/nodes")
+
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC ### Ways
+
+# COMMAND ----------
+
+ways_schema = StructType([
```

Comment:
I thought about that, but I prefer to keep only 3 notebooks for simplicity; so that each one is independently executable without dependencies.

Reply:
Oh, I see that @sllynn did this change already along with a schema refactoring.
For a SW project I would agree 100%, but this is a demo notebook.

I think it is more important to keep each notebook self-contained is more important. If we split it, then it won't work any more in Workspaces (You would need to use Repos). 

Since our goal for this part is to make Mosaic "easy" and not more powerful, then I believe it should be in the same notebook.

Reply:
As per Slack discussion, I am moving this into the Download file in a separate cell.